### DISCUSSION TITLE
Migration from Eslint V8 to V9 using @eslint/migrate-config

### DISCUSSION THREAD

**Discussion by balboFK** [standard/eslint-config-standard-with-typescript#1849] [Q&A] [answered]
Hello everyone!
I am a user of this configuration/package for a long time, since the times of the "standard" name. Now in our project we started the migration from Eslint v8 to v9, and I saw that there was a migrator, so why not give it a shot to see how it goes? Running the migrator in itself produced this file from our original setup: 

```
// eslint.config.mjs

import path from "node:path"
import { fileURLToPath } from "node:url"
import js from "@eslint/js"
import { FlatCompat } from "@eslint/eslintrc"

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const compat = new FlatCompat({
    baseDirectory: __dirname,
    recommendedConfig: js.configs.recommended,
    allConfig: js.configs.all
})

export default [{
    ignores: [
        "**/node_modules",
        "**/dist",
        "**/coverage",
        "./data",
        "**/requirements",
        "**/.vscode",
        "**/jest.config.js",
        "**/jest-integration-config.js",
        "**/jest-mongodb-config.js",
        "**/jest-unit-config.js",
    ],
    files: ["src/**/*.ts", "tests/**/*.ts"],
}, ...compat.extends("love"), {
    languageOptions: {
        ecmaVersion: 5,
        sourceType: "script",
        parserOptions: {
            project: "./tsconfig.json",
        },
    },
    rules: {
        "@typescript-eslint/strict-boolean-expressions": "off",
        "@typescript-eslint/consistent-type-definitions": "off",
        "@typescript-eslint/comma-spacing": "off",
        "@typescript-eslint/return-await": "off",
        "@typescript-eslint/restrict-template-expressions": "off",
        "@typescript-eslint/no-misused-promises": "off",
        "@typescript-eslint/no-namespace": "off",
        "@typescript-eslint/no-confusing-void-expression": "off",
        "@typescript-eslint/key-spacing": ["error", { afterColon: true }],
        "@typescript-eslint/no-extraneous-class": "off",
        "@typescript-eslint/prefer-nullish-coalescing": "off",
        "@typescript-eslint/no-unsafe-argument": "off",
        "key-spacing": "off",
        "no-new-wrappers": "off",
        "import/export": "off",
    },
}]
```

Out of the box it looks ok, but if we stop to pay attention its trying to use "FlatCompat" to set the configuration for "love", and i dont think thats needed since "love" is now compatible with v9 right?

From this config, if we try to lint, we get this error: 

![image](https://github.com/user-attachments/assets/8e728ce4-1252-4c3b-a666-d71ffc56ae56)

I think it might be because of the compat.extends usage?
I was also wondering if it would be easier to migrate this manually than trying to figure out the migrator in itself, and if i could get some help on that. I tried for some hours and didnt manage to make the "old setup" transition to the "new setup". I have numerous trouble with rules breaking, some rules that used to work not working anymore, and some new rules that we never  used, starting to throw errors too (which i think might be expected).

For reference this are my project old configurations:
```
// .eslintrc.json
{
  "extends": "love",
  "parserOptions": {
    "project": "./tsconfig.json"
  },
  "rules": {    
    "@typescript-eslint/strict-boolean-expressions": "off",
    "@typescript-eslint/consistent-type-definitions": "off",
    "@typescript-eslint/comma-spacing": "off",
    "@typescript-eslint/return-await": "off",
    "@typescript-eslint/restrict-template-expressions": "off",
    "@typescript-eslint/no-misused-promises": "off",
    "@typescript-eslint/no-namespace": "off",
    "@typescript-eslint/no-confusing-void-expression": "off",
    "@typescript-eslint/key-spacing": ["error", { "afterColon": true }],
    "@typescript-eslint/no-extraneous-class": "off",   
    "@typescript-eslint/prefer-nullish-coalescing": "off",
    "@typescript-eslint/no-unsafe-argument": "off",
    "key-spacing": "off",
    "no-new-wrappers": "off",
    "import/export": "off"      
  }
}
```
```
// .eslintignore
node_modules
dist
coverage
./data
requirements
.vscode
jest.config.js
jest-integration-config.js
jest-mongodb-config.js
jest-unit-config.js
```

And my tsconfig.json that stayed the same.
```
{
  "compilerOptions": {
    "lib": ["ES2020"],
    "module": "commonjs",
    "target": "ES2020",
    "outDir": "dist",
    "esModuleInterop": true,    
    "sourceMap": true,
    "rootDirs": ["src", "tests"],
    "baseUrl": "src",
    "paths": {
      "@/tests/*": ["../tests/*"],
      "@/*": ["*"]
    }
  },
  "include": ["src", "tests"],
  "exclude": []
}
```

For context this project is a typescript microservice using node + jest for testing.
Any help is welcome.
Thanks!

**Comment by mightyiam** [comment]
Yes, this project is all up-to-date, so no need for that compat utility.

That error seems like you're somehow `require`ing this project, which is not a thing since a few releases ago, where this is now an ES module only package.

Do these help?

**Reply by balboFK** [reply]
@mightyiam  Thanks for changing this into a discussion. Hmm so maybe this might be because of my tsconfig that has it "module" set as "commonjs"? I'll try that to see if this might be the problem, maybe is not on the linting itself, and more on the tsconfig side of things. I have no more JS files on the project since we migrated completely to TS so that config might be an old one that needs updating.

**Reply by mightyiam** [reply]
Yes, that doesn't sound right. But I'm not certain that's the cause of that error.

**Reply by balboFK** [reply]
I went back to this now, and i figured out some things, it seems that the configuration i used with "love", sometimes sets rules on the other parts of the configuration object if i use it directly, so i fixed my tsconfig to be consistent with node 20, and rewrote the whole eslintconfig file to this version: 

```
import globals from "globals"
import pluginJs from "@eslint/js"
import tseslint from "typescript-eslint"
import loveConfig from "eslint-config-love"

export default [
  {
    ignores: [
      "node_modules",
      "dist",
      "coverage",
      "jest.config.js",
      "jest-integration-config.js",
      "jest-mongodb-config.js",
      "jest-unit-config.js",
    ]
  },
  { files: ["**/*.{mjs,ts}"] },
  { languageOptions: { globals: globals.node } },
  pluginJs.configs.recommended,
  ...tseslint.configs.recommended,
  {
    rules: {
      ...loveConfig.rules,
      "no-magic-numbers": "off",
      "key-spacing": "off",
      "no-new-wrappers": "off",
      "import/export": "off",
      "prefer-destructuring": "off",
      "@typescript-eslint/strict-boolean-expressions": "off",
      "@typescript-eslint/consistent-type-definitions": "off",
      "@typescript-eslint/comma-spacing": "off",
      "@typescript-eslint/return-await": "off",
      "@typescript-eslint/restrict-template-expressions": "off",
      "@typescript-eslint/no-misused-promises": "off",
      "@typescript-eslint/no-namespace": "off",
      "@typescript-eslint/no-confusing-void-expression": "off",
      "@typescript-eslint/no-extraneous-class": "off",
      "@typescript-eslint/prefer-nullish-coalescing": "off",
      "@typescript-eslint/no-unsafe-argument": "off",
      "@typescript-eslint/no-import-type-side-effects": "off",
      "@typescript-eslint/no-unnecessary-condition": "off",
      "@typescript-eslint/no-magic-numbers": "off",
      "@typescript-eslint/no-unsafe-assignment": "off",
      "@typescript-eslint/no-explicit-any": "off",      
      "@typescript-eslint/prefer-destructuring": "off",
      "@typescript-eslint/no-unsafe-member-access": "off"
    }
  }
]
```

Since i only wanted the rules from "love" i expected it to work, but seems like some of the dependencies of the "love" configuration fail to work like this and i get this error: 

![image](https://github.com/user-attachments/assets/8be0a5dc-4186-4b07-ae0d-d7a6a4f5ad1e)

I think I'm closer to figuring out the best setup, but still not sure yet with this approach.

**Reply by mightyiam** [reply]
Right. You can't just pick the rules for this project. `configLove` needs to be at the top level array.

**Reply by balboFK** [reply, marked as answer]
Hello again! 
I managed to make it work fine now, i went around digging how this configuration for Eslint worked, stayed a while on their docs and figure out what "love" does for me on the configuration that i did not need to do so i just "attached" the rest of our configuration to it. This is the final file. (theres a lot of rules turned off for now, we are discussing about what to keep off and on).



``` 
// eslint.config.mjs

import loveConfig from "eslint-config-love"

export default [
  loveConfig,  
  {
    ignores: [
      "node_modules",
      "dist",
      "coverage",
      "jest.config.js",
      "jest-integration-config.js",
      "jest-mongodb-config.js",
      "jest-unit-config.js",
      "eslint.config.mjs"
    ]
  },
  { files: ["**/*.{mjs,ts}"] },
  {
    rules: {
      "no-magic-numbers": "off",
      "no-new-wrappers": "off",
      "import/export": "off",
      "prefer-destructuring": "off",
      "no-throw-literal": "off",
      "@typescript-eslint/only-throw-error": "off",
      "@typescript-eslint/strict-boolean-expressions": "off",
      "@typescript-eslint/consistent-type-definitions": "off",
      "@typescript-eslint/return-await": "off",
      "@typescript-eslint/restrict-template-expressions": "off",
      "@typescript-eslint/no-misused-promises": "off",
      "@typescript-eslint/no-namespace": "off",
      "@typescript-eslint/no-confusing-void-expression": "off",
      "@typescript-eslint/no-extraneous-class": "off",
      "@typescript-eslint/prefer-nullish-coalescing": "off",
      "@typescript-eslint/no-unsafe-argument": "off",
      "@typescript-eslint/no-import-type-side-effects": "off",
      "@typescript-eslint/no-unnecessary-condition": "off",
      "@typescript-eslint/no-magic-numbers": "off",
      "@typescript-eslint/no-unsafe-assignment": "off",
      "@typescript-eslint/no-explicit-any": "off",      
      "@typescript-eslint/prefer-destructuring": "off",
      "@typescript-eslint/no-unsafe-member-access": "off",
      "@typescript-eslint/no-unsafe-call": "off",
      "@typescript-eslint/max-params": "off",
      "@typescript-eslint/no-unsafe-return": "off",
      "@typescript-eslint/consistent-type-exports": "off"
    }
  }
]
```

Thanks for your attention @mightyiam, i appreciate it.

**Reply by mightyiam** [reply]
Thank you for sharing, @balboFK .

**Reply by balboFK** [reply]
Also, i think another good thing to share is that most configs might have installed some packages to run their old configs from eslint. But this dependencies are on "love" itself, mainly this ones: 

![image](https://github.com/user-attachments/assets/c2b2b95b-d5c7-4984-a9c3-ea7a828c64d4)


I had then on my package.json too, and if i'm not mistaken i dont think you need it there since "love" already uses it. So it makes the package a little cleaner after the migration. I think this was the case before with eslint v8 too, but its a good clean up to have imo.

**Reply by mightyiam** [reply]
That's correct.
